### PR TITLE
NTP discussion (was mode 6 control packets)

### PR DESCRIPTION
**Pull request by sjvudp**
After about 10 years of frustration with tcpdump decoding NTP control messages as timestamp messages, I decided to try to fix it, thereby implementing NTP mode 6 Control Message processing.
I did not do the data portion yet, but this is already quite useful. Along with the new parsing came a little renaming of strings. I tried to make commits for logical steps.
This is my very first github effort, so please be forgiving.

### CODE REVIEW

**Comment by infrastation** [comment]
The CI check has failed because the code tries to print the timestamps in local time rather than UTC, hence the test only passes when the local time offset is the same. This needs to be fixed. Otherwise it looks like an organized set of changes to it should be easier to read when somebody gets to examine it closer.

**Comment by sjvudp** [comment]
The timezone bug is old code; I just added a test that fails now; I'll have a look...

>>> Denis Ovsienko <notifications@github.com> schrieb am 23.08.2017 um 16:21 in
Nachricht <the-tcpdump-group/tcpdump/pull/625/c324349238@github.com>:

**Comment by sjvudp** [comment]
I fixed the `localtime()` time stamps, converting them to `gmtime()` time stamps, thereby also touching up rounding. You can diff the test output to see then change, as that was required to be updated anyway.

**Comment by infrastation** [comment]
Thank you very much, please let us have some time to make a more thorough review.

**Comment by sjvudp** [comment]
I added some more work for NTP processing, adding some details, streamlining some details...
Unfortunately I had to update the test results again, as the output is somewhat different now.
Specifically I ran each of the two tests at four verbosity levels (0-3).

**Comment by sjvudp** [comment]
I tried to finish the task of mode 6 (control message) packet decoding by decoding the different status words and their components. Unfortunately the output format was changed by doing so. However I hope the format is more consistent now. The depth of indent is discussable, maybe...

**Comment by infrastation** [comment]
There is pull request #607 in the way of this pull request (the changes don't merge well), I will try to minimize that pull request first and let you know how it goes.

**Comment by infrastation** [comment]
Could you rebase on the current master branch and place simpler changes first if it can be done easily? Thank you.

**Comment by sjvudp** [comment]
I integrated the fix for unsigned Poll (a rather academic case as there are no valid negative polling intervals) and the printing of the Reference IDs (which is rather cosmetic). I did not integrate the fix for the digest size, because I don't understand it: The digest is 128 bits, which makes 16 octets, not 20.
If you tell me what "simpler fixes" are, I (or maybe someone else) might rebase them on master.

**Comment by sjvudp** [comment]
Resolving the conflicts: As I integrated the fixes of mlichvar (somewhat more elegantly as I believe), I suggest to integrate my branch at commit d1a2d1853986740219664fd009b8eaa0e0416758, because that's where I forked it. There should be no loss of fixed functionality.

**Comment by sjvudp** [comment]
Integrated the fix for printing NTP's KeyID with the correct endian-ness, pulled upstream, then pushed the new candidate.

File: tests/TESTLIST [hunk 1]
```diff
@@ -310,7 +310,7 @@ dhcp-rfc3004	dhcp-rfc3004.pcap	dhcp-rfc3004-v.out	-v
 dhcp-rfc5859	dhcp-rfc5859.pcap	dhcp-rfc5859-v.out	-v
 dhcp-mud	dhcp-mud.pcap		dhcp-mud.out	-vv
 
-# MEDSA tests
+# MEDSA (NTP?) tests
```
Comment (infrastation) [inline tests/TESTLIST]: This is a sample of Marvell EDSA link type decoding, it just happens to contain IP/UDP/NTP among other packets.

**Comment by sjvudp** [comment]
>>> Denis Ovsienko <notifications@github.com> schrieb am 29.08.2017 um 15:35 in
Nachricht <the-tcpdump-group/tcpdump/pull/625/review/59241430@github.com>:
> infrastation commented on this pull request.

> -# MEDSA tests
> +# MEDSA (NTP?) tests
> 
> This is a sample of Marvell EDSA link type decoding, it just happens to 
> contain IP/UDP/NTP among other packets.

So wouldn't it make sense to filter out the NTP packets from the example, then?
As a side-effect reducing the size of the test input...

**Comment by infrastation** [comment]
Keeping the MEDSA test case as it is is OK, protocols get nested into each other quite often.

Generally this looks like a useful work, let me suggest how to integrate it best.

4 trivial commits from this pull request are now in the master branch. If you rebase your branch on the master branch those 4 will cancel out and make the pull request slightly simpler, also the history will become linear again. Also the trivial changes from the earlier pull request are in the master branch so they will not normally show in what you want to add on top of the master branch.

Besides that, it will help to separate the work that you think is 100% complete from the work that is still in progress because it is difficult to review everything at once, especially when new commits come in once in a few days. Perhaps a volatile development branch and a branch of complete and tested changes would work better.

**Comment by sjvudp** [comment]
As said initially, this is my first attempt for a github contribution. For my part I tried to deliver the feature in logical chunks that are for themself complete. That's why I used non-fast-forward merges. You could reorganize my merges into one single commit if you like. I guess it's my non-ff merges that let you talk about non-linear history. I wonder whether I should start a new branch that rebases my patch collection and abandon this pull request (I don't know how github will handle all the seemingly new commits then). I don't know when I have time to continue to work on this branch, but in principle interpretation of the data part of the control messages is planned for thus feature branch. As long as noone else works on it (NTP Control Messages), it should be free of conflicts. I hope ;-)

**Comment by sjvudp** [comment]
I tried to resolve conflicts on **pull request 627**, but there are still conflicts indicated. I prepared another local branch (rebased this time), but there are also conflicts indicated, so I did not add a pull request. I need some help, please!

**Comment by infrastation** [comment]
Please leave both your pull requests as they are now, I am trying to simplify the history of your work, will update shortly.

**Comment by sjvudp** [comment]
> Please leave both your pull requests as they are now, I am trying to simplify  the history of your work, will update shortly.

Maybe look at the rebased branch at f-NTP-mode6-B3...

**Comment by infrastation** [comment]
I have tried to rebase this work and could not do it quickly because some commits mix changes of different kind and when a merge conflict happens it is difficult to understand why lots of code fails to be replaced with lots of other code. There is an added length check for NTP in the master branch, it adds to the complexity of rebase a bit but I believe it helps to avoid problems in future.

If you can afford more time I would suggest to do as follows. As soon as this is the oldest NTP-related pull request in the queue now, if you continue to work on it you can be reasonably sure subsequent pull requests will not get in the way of your work (we aim to process requests FIFO where possible). Also if you don't add new work to the set of commits it will not become more difficult.

If you split the changes into smaller, more focused and isolated changes, it will be easier to review and merge, and also to troubleshoot later if necessary. For example, if you see that pre-existing whitespace is wrong, make necessary whitespace fixes (and those fixes only) before and separately from the main work, get them merged and forget about it. Alternatively, if you don't care about pre-existing whitespace, leave it as it is. In either case, get whitespace right in your commits.

The very first commit that makes it difficult is "print-ntp.c: Handle NTP Control Messages", can you make it simpler? It turns out renaming `bp` to `td` introduces many differences that also shows in the rebase later. However, it does not influence the code path so I would suggest not to rename it if possible. Also `ntp_control_data` needs to use `nd_uint8_t` and `nd_uint16_t` types to avoid potential alignment troubles (the input to the printer function is a sequence of bytes, if it does not happen to be aligned with the structure it will cause bugs sooner or later). On that note, existing structures in `print-ntp.c` will benefit from using those types too, feel free to fix them yourself (as a separate commit) before the main changes.

To edit the commit without getting into git details too deep you can create a new branch from the most recent upstream master branch, `git checkout` it to make it your working copy and `git cherry-pick` the commit you need to edit. It will conflict because of the new length checks, edit the file to go with the new checks and once happy run `git add` and `git commit` (this will finalize the cherry-pick and your working branch will be 1 commit ahead of master). After adding the changes you want run `git add` and 'git commit --amend`, this will incorporate them into the last commit without creating a new one (still 1 commit ahead). The point is, in the end you should have a commit that only changes the code when necessary and only does one thing.

After you have some results that make a complete change (that is, safe to be left unattended after the merge) and don't conflict with the master branch, we will try to review and merge, making the total amount of unmerged work smaller after each iteration. Will it work for you?

**Comment by infrastation** [comment]
By the way, the NTP test cases from this set of commits can be added before any other changes, subsequent commits would just have to include the changes they make to the .out files.

**Comment by mcr** [comment]
sjvudp <notifications@github.com> wrote:
    > part I tried to deliver the feature in logical chunks that are for themself
    > complete. That's why I used non-fast-forward merges. You could reorganize my
    > merges into one single commit if you like. I guess it's my non-ff merges that
    > let you talk about non-linear history. I wonder whether I should start a new
    > branch that rebases my patch collection and abandon this pull request
    > (I

Yes, git fetch, git rebase -i origin/master
It's often that simple.

--
]               Never tell me the odds!                 | ipv6 mesh networks [
]   Michael Richardson, Sandelman Software Works        | network architect  [
]     mcr@sandelman.ca  http://www.sandelman.ca/        |   ruby on rails    [

**Comment by sjvudp** [comment]
Infrastation, thanks for the comments; a few of my thoughts follow:
I'm unsure which "length check" you are exactly referring to. Probable it's better to compare the whole function instead of diff chunks between master and my branch. Some changes need restructuring of existing code.
On big diffs and white spaces: I thought you prefer rather complete sets of changes, avoiding too many small changes. It looks like i was wrong. As I'm using Emacs, some of the space fixes happen automatically. Sometimes I'll try to make future diffs more clean, so usually I put printf-like format strings and their parameters on different lines, so then it's easier to see whether the change is on format or on value. And at other times I just fold lines that are too long (and should never be there).

On the "Handle NTP Control Messages" diff: What I could have separated is normalizing LI. The rest is almost the minimal restructuring change needed to continue.

On renaming of "bp": In the old ntp_print() there was a "cp" eventually cast to "bp" being a pointer to struct ntpdata. My replacement ntp_time_print() receives "td" (for "time data") as a pointer to struct ntp_time_data. Obviously "NTP data" comes in two completely different packet formats, so I added union ntpdata. For compatibility I had retained the old names "cp" and "bp" in the new ntp_print().

On nd_uint8_t and nd_uint16_t: Thanks, I didn't know that.

Personally I'll try to complete what I intended for this branch locally (continuing on the original branch that was fine until you merged earlier stuff), and then I might add another branch on top of upstream master to receive new changes.

On the testcase: That is rather quick and dirty; I'll try to produce a better one.

**Comment by infrastation** [comment]
Let me know if you get stuck. Would merging #628 get in the way of your work?

**Comment by sjvudp** [comment]
It all depends: In addition to the tests provided (quickly and dirty), I have two more test files here, and I did some interactive testing during recent development. I don't know a mechanism to "cherry pick" single packets from a pcap file, or a tool that will fix UDP checksums after editing some packets (in case desired packets are hard to capture, but anyway several kinds of different packets are desired for testing: ordinary time packets (hard to get those from old versions these days) with and without MAC, different types of MACs, successful and failed mode 6 packets, seemingly invalid or truncated packets, etc.
And let me remark what I did locally so far: I completed the interpretation of the data portion of mode 6 requests and responses (which is quite nice, I think). The output format and indent was streamlines once again (I don't like to do this often, but the result is quite good now). The final touch-up will be interpretation of the padding and MAC parts, and maybe do a basic handling of mode 7 requests/responses.
When done and satisfied (talking about myself), I'll try to make small chunks of fixes on upstream master. Which rebase I'm going to about the same work, I'm afraid, but it won't provide the small logical commits you like to see. So I intend to add interactively added chunks to an integration branch, and finally push the result to some branch.

**Comment by sjvudp** [comment]
Just for demonstration: A preview of the current state of mode 6 decoding:
```
15:16:18.641347 IP6 (flowlabel 0xb498e, hlim 64, next-header UDP (17) payload length: 60) localhost.54510 > localhost.ntp: [bad udp cksum 0x004f -> 0x08e6!] NTP leap indicator=0 (Nominal), Version=2, Mode=6 (Control Message), length=52
    R=Request, E=OK, M=Last, OpCode=10 (retrieve remote stats), Sequence=31
    Status=0000, Assoc.=0, Offset=0, Count=40, Status=0000
    data:
        nonce=dd53dc22a41bb28510547f3d, frags=32
15:16:18.641548 IP6 (class 0xb8, flowlabel 0xa17fc, hlim 64, next-header UDP (17) payload length: 256) localhost.ntp > localhost.54510: [bad udp cksum 0x0113 -> 0x167d!] NTP leap indicator=0 (Nominal), Version=2, Mode=6 (Control Message), length=248
    R=Response, E=OK, M=Last, OpCode=10 (retrieve remote stats), Sequence=31
    Status=0000, Assoc.=0, Offset=0, Count=233, Status=0000
    data:
        nonce=dd53dc22a42f5e8c5d2ef23f, rs.0=0x1d0, first.0=0xdd3fc359.a354e70f,^M^Jct.0=460, mv.0=36, last.0=0xdd53d8b6.8b33be5c, addr.0=132.199.4.1:123,^M^J^SH.0=18446744073709487267, now=0xdd53dc22.a43906ed,^M^Jlast.newest=0xdd53d8b6.8b33be5c^M^J
```

**Comment by infrastation** [comment]
This is a good potential contribution overall, just the details keep getting in the way, so let's see how we can make it. Can you prepare the first complete pending commit (print-ntp.c: Handle NTP Control Messages) on top of the current master branch the way you consider the best? If that is difficult because of the merge conflicts I can do that for you so you can buffer subsequent changes on top of that.

**Comment by sjvudp** [comment]
Pushed a new branch i-master that has a minimalistic subset of the first three commits of this branch. So I'll have the rebasing fun ;-)

**Comment by infrastation** [comment]
Are you happy with that commit to go into master?

**Comment by sjvudp** [comment]
For me there are no objections, but I wonder what to do with this pull request: It seems I cannot "reset" it. Should I _start another pull request_ and abandon _this one_?

**Comment by infrastation** [comment]
Thank you, the NTP control message support commit is now in the master branch. To merge it it was necessary to resolve a merge conflict because your branch was not up to date with master, I have done that. Also the commit message repeated your commits that were already in the master branch, I have left that out. Also there were a few code issues: one about code style and the others about code stability and security -- those had to be done immediately after your change. All that is sorted now, please make a copy of the current master branch and prepare the next complete commit or set of commits on top of that.

The pull request, you can make a forced push into its git branch and it will replace the commits that show up on GitHub. Of course, this can easily destroy your work in progress if you don't do it carefully, so if you don't want to experiment just leave the pull request as it is for now and close it after all your work is in.

Let me know if you have any questions or the next change to discuss.

**Comment by sjvudp** [comment]
As it turned out, I pulled the upstream master to `8dda56aa`, but something went wrong, and my integration branch was still based on `bd3f9392`. I'm sorry for that!
Trying to find out what you did to correct the issues...

**Comment by sjvudp** [comment]
Here's my work that mostly completes the mode 6 processing: I've added decoding the data portions, the padding and the message digest. Found that reality does not always match the specs. Example for a failed authentication of a mode 6 control message:
```
13:56:33.985442 IP6 (flowlabel 0x053b7, hlim 64, next-header UDP (17) payload length: 44) ::1.46618 > ::1.123: [bad udp cksum 0x003f -> 0xa40d!] NTP leap indicator=0 (Nominal), Version=2, Mode=6 (Control Message), length=36
    R=Request, E=OK, M=Last, OpCode=11 (retrieve ordered list), Sequence=36
    Status=0000, Assoc.=0, Offset=0, Count=0, PadLen=4
    MAC:
        KeyID=2, Digest=a557d7e8 aa926d99 c7e37a53 2cb88a64
13:56:33.985596 IP6 (class 0xb8, flowlabel 0x9235b, hlim 64, next-header UDP (17) payload length: 20) ::1.123 > ::1.46618: [bad udp cksum 0x0027 -> 0x3140!] NTP leap indicator=0 (Nominal), Version=2, Mode=6 (Control Message), length=12
    R=Response, E=Error, M=Last, OpCode=11 (retrieve ordered list), Sequence=36
    Status=0x100, Assoc.=0, Offset=0, Count=0
        ErrStat=0x100 (Code=1 (auth. failure), Reserved=0)
```

**Comment by mcr** [comment]
> For me there are no objections, but I wonder what to do with this pull
> request: It seems I cannot "reset" it. Should I start another pull request
> and abandon this one?

If you push a new branch with the same name to github, it will reset.
"git push -f origin mybranchname"




--
]               Never tell me the odds!                 | ipv6 mesh networks [
]   Michael Richardson, Sandelman Software Works        | network architect  [
]     mcr@sandelman.ca  http://www.sandelman.ca/        |   ruby on rails    [

**Comment by sjvudp** [comment]
Just added a minor improvement for MAC which looks like this now:
```
    MAC(size 20):
        KeyID=2, MD5-Digest=a557d7e8 aa926d99 c7e37a53 2cb88a64
```

**Comment by infrastation** [comment]
OK, what is the next complete change ready to go into the master branch and where is it?

**Comment by sjvudp** [comment]
I have a pushed rebased branch named i`-NTP-mode-6`. How to proceed?

**Comment by infrastation** [comment]
The branch i-NTP-mode-6 is based on the master branch as it was 2 weeks ago (20 August 2017) and it complicates the history of changes since that point because it forks, merges and reorders commits. Could you please make a new temporary branch that is an exact copy of the master branch of the repository the-tcpdump-group/tcpdump as it is right now? After you decide which commit(s) should go into the master branch first please cherry-pick those commit(s) only to that temporary branch without fork/merge commits. If you could do that it would help a lot to reduce the amount of unmerged work.

Alternatively, I can cherry-pick the commits from that branch into master for you if you tell which particular commits should go first.

**Comment by infrastation** [comment]
Update: if you make a new temporary branch please wait until tomorrow as we may have lots of changes in the master branch today, so you don't waste your time doing this twice.

**Comment by sjvudp** [comment]
I don't really understand what went wrong: I did a `pull` and a `fetch` on upstream `master` before I did the `rebase`. At the moment I have `master` at `bd3f9392 Print NTP MAC with 160-bit digest.`.
My `upstream` is [https://github.com/the-tcpdump-group/tcpdump](https://github.com/the-tcpdump-group/tcpdump).

**Comment by infrastation** [comment]
Upstream master right now is at commit 80deba1 (NTP: Add missing bounds checks.), do you see the same when you follow the link you have given?

**Comment by sjvudp** [comment]
Verry odd: I fetched and pulled, but the latest log item is still `commit bd3f9392a61`. However when I freshly clone the repository using the same URL, the latest log entry in master is `commit 80deba15`.

In my working repository I set up remotes like this:
```
git remote -v
local1	../tcpdump.orig (fetch)
local1	../tcpdump.orig (push)
mlichvar	https://github.com/mlichvar/tcpdump.git (fetch)
mlichvar	https://github.com/mlichvar/tcpdump.git (push)
origin	https://github.com/sjvudp/tcpdump (fetch)
origin	https://github.com/sjvudp/tcpdump (push)
upstream	https://github.com/the-tcpdump-group/tcpdump (fetch)
upstream	https://github.com/the-tcpdump-group/tcpdump (push)
```

And on the test repository I have:
```
git remote -v
origin	https://github.com/the-tcpdump-group/tcpdump (fetch)
origin	https://github.com/the-tcpdump-group/tcpdump (push)
```
So my understanding was that fetch upstream would fetch the changes into my working repository. I'm clueless regarding this problem, sorry. git version 2.12.3.

**Comment by infrastation** [comment]
`git fetch` updates the history recorded about the remote branches, it does not merge that remote history into local branches. `git pull` does that as it is a wrapper around `git fetch` and `git merge`. This requires some time to master, meanwhile you may find it more convenient to clone a fresh temporary repository after the upstream changes, then to put the intended changes into that temporary repository, push to GitHub from there and request to merge. After everything you want is in the master branch you will not need the temporary copies anymore.

Let me know once you have the next changes ready.

**Comment by sjvudp** [comment]
I spent **a lot** of time for a new branch `i2-NTP-mode6` that rebases these patches on `commit eb025532`.
Most of the extra effort was caused by the fact that two pull requests handle the same problems, and the smaller change set created merge conflicts for the larger change set (mine). I'd wish before starting to merge such things on master, one should talk to the authors whether one can live with the way the other one has fixed issues. While the fixes may seem to star tdifferently, they are quite similar at the end, but git's rebase doesn't know, and I have to pull the incompatible changes quite a long distance...

**Comment by infrastation** [comment]
Thank you, I have just looked through, seen progress and left some comments.

**Comment by sjvudp** [comment]
Sorry, newbie: Where will I find those comments?

**Comment by infrastation** [comment]
If you edit pull request #627 to use your latest branch I will be able to simplify the history in the branch a little bit.

**Comment by sjvudp** [comment]
> If you edit pull request #627 to use your latest branch (...)
How would I do that (I'd like to keep the original (this) branch) at least for a while? Should I swap branch names by doing a triple-rename of branches and then do a forced push ,or what?

**Comment by infrastation** [comment]
Excuse me, I was wrong. Just checked and in the pull request properties it is possible to change the destination branch of the pull request but not the source. If you just open another pull request from i2-NTP-mode6 it should do the job.

**Comment by sjvudp** [comment]
`If you just open another pull request from i2-NTP-mode6 it should do the job.`

Done (#631)

**Comment by infrastation** [comment]
Thank you, I need a couple more days to look at it unless anybody else is willing.

**Comment by sjvudp** [comment]
I decided to continue with a new pull request; maybe just because I don't know how to switch the branch of this pull request to another (rebased) branch. See pull request 657.

**Pull request closed without merging**